inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp

```diff
@@ -126,480 +153,750 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
+                                    int* detections,
+                                    int* indicesOut,
+                                    const float* bboxes,
+                                    const float* sizes) {
+    // Input is candidate for image, output is candidate for each class within image
+    int countIn = detections[0];
+    detections[0] = 0;
+
+    for (int i = 0; i < countIn; ++i) {
+        const int idx = indicesIn[i];
+        const int cls = idx / priorsNum;
+        const int prior = idx % priorsNum;
+
+        // nms within this class
         int &ndetection = detections[cls];
-        int *pindices = indices + cls*_num_priors;
+        int *pindices = indicesOut + cls * priorsNum;
 
         bool keep = true;
         for (int k = 0; k < ndetection; ++k) {
-            const int kept_idx = pindices[k];
+            const int keptPrior = pindices[k];
             float overlap = 0.0f;
-            if (_share_location) {
-                overlap = JaccardOverlap(bboxes, sizes, prior, kept_idx);
+            if (isShareLoc) {
+                overlap = JaccardOverlap(bboxes, sizes, prior, keptPrior);
             } else {
-                overlap = JaccardOverlap(bboxes, sizes, cls*_num_priors + prior, cls*_num_priors + kept_idx);
+                overlap = JaccardOverlap(bboxes, sizes, cls * priorsNum + prior, cls * priorsNum + keptPrior);
             }
-            if (overlap > _nms_threshold) {
+            if (overlap > NMSThreshold) {
                 keep = false;
                 break;
             }
         }
+
         if (keep) {
             pindices[ndetection++] = prior;
         }
     }
 }
 
+void MKLDNNDetectionOutputNode::generateOutput(float* reorderedConfData, int* indicesData, int* detectionsData, float* decodedBboxesData, float* dstData) {
+    const int numResults = getChildEdgesAtPort(0)[0]->getDims()[2];
+    const int DETECTION_SIZE = getChildEdgesAtPort(0)[0]->getDims()[3];
+    if (DETECTION_SIZE != 7) {
+        IE_THROW() << errorPrefix << NOT_IMPLEMENTED;
+    }
+
+    int dstDataSize = 0;
+    if (keepTopK > 0)
+        dstDataSize = imgNum * keepTopK * DETECTION_SIZE * sizeof(float);
+    else if (topK > 0)
+        dstDataSize = imgNum * topK * classesNum * DETECTION_SIZE * sizeof(float);
+    else
+        dstDataSize = imgNum * classesNum * priorsNum * DETECTION_SIZE * sizeof(float);
+
+    if (dstDataSize > getChildEdgesAtPort(0)[0]->getBlob()->byteSize()) {
+        IE_THROW() << errorPrefix << OUT_OF_BOUNDS;
+    }
+    memset(dstData, 0, dstDataSize);
+
+    // set final detection result to output blob
+    int count = 0;
+    for (int n = 0; n < imgNum; ++n) {
+        const float *pconf   = reorderedConfData + n * confInfoLen * classesNum;
+        const float *pboxes  = decodedBboxesData + n * priorsNum * 4 * locNumForClasses;
+        const int *pindices  = indicesData + n * classesNum * priorsNum;
+
+        for (int c = 0; c < classesNum; ++c) {
```

Comment:
Could we use parallel_2d?

Reply:
The output order matters. We can not paralleled directly. We can parallel by computing the output position offset of each batch and class with traverse all detection outputs, but this add additional overhead, together with this part consuming very small proportion times compare overall time of DO layer, i.e. not heavy workload. so does not give obvious perf benefit in most case. I think we can keep it to make it easy to understand. what do you think. Thank you.